### PR TITLE
[RFC DO NOT MERGE] internal fsmonitor

### PR DESCRIPTION
**Pull request by kewillford**
This is the code for an internal fsmonitor.

### CODE REVIEW

File: builtin/fsmonitor--daemon.c [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+struct ipc_data {
+	struct ipc_command_listener data;
+	struct fsmonitor_daemon_state *state;
+};
+
+static int handle_client(struct ipc_command_listener *data,
+			 const char *command,
+			 int (*reply)(void *reply_data,
+				      const char *message, size_t len),
```
Comment (derrickstolee) [inline builtin/fsmonitor--daemon.c]: It would be good to create a `typedef` for this function pointer type.

File: builtin/fsmonitor--daemon.c [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+	struct strbuf token = STRBUF_INIT;
+	intmax_t count = 0;
+
+	trace2_data_string("fsmonitor", the_repository, "command", command);
+
+	strbuf_addf(&token, "%"PRIu64"", state->latest_update);
+	if (!strcmp(command, "quit")) {
+		if (fsmonitor_listen_stop(state))
+			error("Could not terminate watcher thread");
+		sleep_millisec(50);
```
Comment (derrickstolee) [inline builtin/fsmonitor--daemon.c:133]: hm...

File: t/t0052-simple-ipc.sh [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+test_description='simple command server'
+
+. ./test-lib.sh
+
+test_lazy_prereq SUPPORTS_SIMPLE_IPC '
+	test-tool simple-ipc SUPPORTS_SIMPLE_IPC
+'
+
+test_expect_success SUPPORTS_SIMPLE_IPC 'simple command server' '
+	{ test-tool simple-ipc daemon & } &&
```
Comment (derrickstolee) [inline t/t0052-simple-ipc.sh:20]: Should we track the process id and kill it when the test is over?

**Review by derrickstolee — COMMENTED** [review]
Just a few things that popped out to me during an initial scan.

File: t/t0052-simple-ipc.sh (hunk 3 above)
Comment (dscho) [reply t/t0052-simple-ipc.sh:20]: Good idea!

File: builtin/fsmonitor--daemon.c (hunk 1 above)
Comment (dscho) [reply builtin/fsmonitor--daemon.c]: Done: https://github.com/microsoft/git/pull/256/commits/00a2f904200edd076e2f03bbacaf2ddff2e03107#diff-c0c002cce014456cab8d633d56c721eaR35

File: t/t0052-simple-ipc.sh (hunk 3 above)
Comment (dscho) [reply t/t0052-simple-ipc.sh:20]: Done (finally!): https://github.com/microsoft/git/blob/45a31e825ce8a1c00a506db4b18a0fb9d4d8e3f7/t/t0052-simple-ipc.sh#L12-L23

**Comment by dscho** [comment]
@jeffhostetler I integrated your patches to help make `pkt-line` thread-safe (I saw your comment regarding the static `packet_buffer` and I spent about half an hour to do away with it, but the result is too ugly to live, really.

There might be a bit of contention around 4af4f571c1b8c5dc6c9264b70065e299c49e67bb: the function `packet_write_fmt_1()` is called quite often, and the entire point of having that `strbuf` as a `static` is to avoid frequent memory allocations/releases.

On the other hand, 8abec7eddeba32418eff504c6ead9c1a9003736a is not only trivially correct, it is also _quite_ necessary in the context of the FSMonitor, as we really want to allow concurrent client threads (which all read/write packets).

So: thank you for bringing those patches to the table!

**Comment by dscho** [comment]
Closing, in favor of https://github.com/microsoft/git/pull/273.

**Pull request closed without merging**